## Preprocessing

In [4]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
import os

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [6]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN', 'NAME'])

# Drop the 'ASK_AMT' column
application_df = application_df.drop(columns=['ASK_AMT'])

# Display the first few rows of the dataframe to confirm the drop
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  IS_SUCCESSFUL  
0   Association       1              0                      N              1  
1  Co-operative       1         1-9999                      N              1  
2   Association       1              0                      N              0  
3         Trust       1    10000-24999                      N              1  
4         Trust       1  100000-499999                      N              1

In [8]:
# Determine the number of unique values in each column.
unique_values = application_df.nunique()
# Display the unique values for each column
unique_values

APPLICATION_TYPE          17
AFFILIATION                6
CLASSIFICATION            71
USE_CASE                   5
ORGANIZATION               4
STATUS                     2
INCOME_AMT                 9
SPECIAL_CONSIDERATIONS     2
IS_SUCCESSFUL              2
dtype: int64

In [10]:
 # Determine the number of unique values in each column.
application_df.nunique()


APPLICATION_TYPE          17
AFFILIATION                6
CLASSIFICATION            71
USE_CASE                   5
ORGANIZATION               4
STATUS                     2
INCOME_AMT                 9
SPECIAL_CONSIDERATIONS     2
IS_SUCCESSFUL              2
dtype: int64

In [12]:
# Look at APPLICATION_TYPE value counts to identify and replace with "Other"
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
print(application_type_counts)

APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64


In [14]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
cutoff = 528
application_types_to_replace = application_df['APPLICATION_TYPE'].value_counts()
application_types_to_replace = application_types_to_replace[application_types_to_replace < cutoff].index.tolist()

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [16]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"
classification_counts = application_df['CLASSIFICATION'].value_counts()
print(classification_counts)

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64


In [18]:
# You may find it helpful to look at CLASSIFICATION value counts >1
classification_counts_more_than_one = classification_counts[classification_counts > 1]
print(classification_counts_more_than_one)

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: count, dtype: int64


In [20]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
cutoff = 1883 
classifications_to_replace = classification_counts[classification_counts < cutoff].index.tolist()


# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure replacement was successful
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: count, dtype: int64

In [22]:
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  IS_SUCCESSFUL  
0   Association       1              0                      N              1  
1  Co-operative       1         1-9999                      N              1  
2   Association       1              0                      N              0  
3         Trust       1    10000-24999                      N              1  
4         Trust       1  100000-499999                      N              1

In [24]:
# Convert categorical data to numeric with `pd.get_dummies`
# Convert categorical data to numeric with `pd.get_dummies`
application_df_dummies = pd.get_dummies(application_df)
# Convert Boolean columns to integers (1 and 0)
application_df_dummies = application_df_dummies * 1

# Display the first few rows of the new DataFrame to verify the conversio
application_df_dummies.head()

STATUS  IS_SUCCESSFUL  APPLICATION_TYPE_Other  APPLICATION_TYPE_T10  \
0       1              1                       0                     1   
1       1              1                       0                     0   
2       1              0                       0                     0   
3       1              1                       0                     0   
4       1              1                       0                     0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  \
0                     0                    0                    0   
1                     0                    1                    0   
2                     0                    0                    0   
3                     0                    1                    0   
4                     0                    1                    0   

   APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  ...  \
0                    0                    0                    0  ...   
1                    0                    0                    0  ...   
2                    1                    0                    0  ...   
3                    0                    0                    0  ...   
4                    0                    0                    0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0                0                  0                         1   
1                0                  0                         1   
2                0                  0                         1   
3                0                  0                         1   
4                0                  0                         1   

   SPECIAL_CONSIDERATIONS_Y  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  

[5 rows x 43 columns]

In [26]:
# Split our preprocessed data into our features and target arrays
y = application_df_dummies['IS_SUCCESSFUL'].values
X = application_df_dummies.drop(columns='IS_SUCCESSFUL').values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [28]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [180]:
# Second attempt- 72.76%
# Set the random seed
import numpy as np
import random
seed = 42
np.random.seed(seed)
tf.random.set_seed(seed)
random.seed(seed)
# Define the model - deep neural net
nn = tf.keras.models.Sequential()

# Input layer and first hidden layer
nn.add(tf.keras.layers.Dense(units=80, input_dim=X_train.shape[1], activation='relu'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=30, activation='relu'))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=15, activation='relu'))

# Dropout layer for regularization
nn.add(tf.keras.layers.Dropout(0.2))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_6"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_27 (Dense)                │ (None, 80)             │         3,440 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_28 (Dense)                │ (None, 30)             │         2,430 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_29 (Dense)                │ (None, 15)             │           465 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_6 (Dropout)             │ (None, 15)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_30 (Dense)                │ (None, 1)              │            16 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 6,351 (24.81 KB)

 Trainable params: 6,351 (24.81 KB)

 Non-trainable params: 0 (0.00 B)

In [181]:
# Compile the model
nn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
# Check the structure of the model
nn.summary()

Model: "sequential_6"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_27 (Dense)                │ (None, 80)             │         3,440 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_28 (Dense)                │ (None, 30)             │         2,430 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_29 (Dense)                │ (None, 15)             │           465 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_6 (Dropout)             │ (None, 15)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_30 (Dense)                │ (None, 1)              │            16 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 6,351 (24.81 KB)

 Trainable params: 6,351 (24.81 KB)

 Non-trainable params: 0 (0.00 B)

In [182]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=150, batch_size=32, verbose=1, validation_split=0.2)


Epoch 1/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 1s 577us/step - accuracy: 0.6932 - loss: 0.5988 - val_accuracy: 0.7345 - val_loss: 0.5513
Epoch 2/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 403us/step - accuracy: 0.7237 - loss: 0.5622 - val_accuracy: 0.7361 - val_loss: 0.5485
Epoch 3/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 417us/step - accuracy: 0.7285 - loss: 0.5614 - val_accuracy: 0.7361 - val_loss: 0.5483
Epoch 4/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 385us/step - accuracy: 0.7250 - loss: 0.5586 - val_accuracy: 0.7353 - val_loss: 0.5480
Epoch 5/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 369us/step - accuracy: 0.7286 - loss: 0.5557 - val_accuracy: 0.7300 - val_loss: 0.5476
Epoch 6/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 351us/step - accuracy: 0.7270 - loss: 0.5539 - val_accuracy: 0.7312 - val_loss: 0.5475
Epoch 7/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 355us/step - accuracy: 0.7264 - loss: 0.5534 - val_accuracy: 0.7357 - val_loss: 0.5467
Epoch 8/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 352us/step - accuracy: 0.7291 - loss: 0

In [183]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - 390us/step - accuracy: 0.7276 - loss: 0.5802
Loss: 0.5802252292633057, Accuracy: 0.727580189704895


# Attempt 2 Results

** Accuracy:

The accuracy slightly improved compared to the previous version with ASK_AMT transformed (72.01%) and is higher than the earlier seeded version without ASK_AMT changes (72.71%).
This suggests that removing ASK_AMT streamlined the model without negatively impacting its predictive power.

** Loss:

The loss is slightly higher (0.5802) compared to the seeded version without ASK_AMT (0.5698) but remains reasonable.
Impact of Dropping ASK_AMT:

Removing ASK_AMT reduces potential noise or redundancy, leading to a slight improvement in accuracy for this attempt.

In [185]:
# Export our model to HDF5 file
nn.save("H5/AlphabetSoupCharity_Attempt_2.h5")

print("Model saved successfully!")

Model saved successfully!


In [30]:
# Third Attempt - 72.83% accuracy
# Set the random seed
import numpy as np
import random
seed = 42
np.random.seed(seed)
tf.random.set_seed(seed)
random.seed(seed)

# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
nn = tf.keras.models.Sequential()

# Input layer and first hidden layer
nn.add(tf.keras.layers.Dense(units=100, input_dim=X_train.shape[1], activation='relu'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=50, activation='relu'))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=30, activation='relu'))

# Fourth hidden layer
nn.add(tf.keras.layers.Dense(units=15, activation='relu'))

# Dropout layer
nn.add(tf.keras.layers.Dropout(0.1))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()


/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 100)            │         4,300 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 50)             │         5,050 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 30)             │         1,530 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 15)             │           465 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 15)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 1)              │            16 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 11,361 (44.38 KB)

 Trainable params: 11,361 (44.38 KB)

 Non-trainable params: 0 (0.00 B)

In [32]:
# Compile the model
nn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
# Check the structure of the model
nn.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 100)            │         4,300 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 50)             │         5,050 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 30)             │         1,530 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 15)             │           465 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 15)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 1)              │            16 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 11,361 (44.38 KB)

 Trainable params: 11,361 (44.38 KB)

 Non-trainable params: 0 (0.00 B)

In [34]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=150, batch_size=32, verbose=1, validation_split=0.2)


Epoch 1/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 1s 470us/step - accuracy: 0.6944 - loss: 0.6010 - val_accuracy: 0.7326 - val_loss: 0.5530
Epoch 2/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 390us/step - accuracy: 0.7242 - loss: 0.5618 - val_accuracy: 0.7318 - val_loss: 0.5505
Epoch 3/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 416us/step - accuracy: 0.7256 - loss: 0.5594 - val_accuracy: 0.7345 - val_loss: 0.5484
Epoch 4/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 384us/step - accuracy: 0.7283 - loss: 0.5550 - val_accuracy: 0.7364 - val_loss: 0.5467
Epoch 5/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 382us/step - accuracy: 0.7270 - loss: 0.5536 - val_accuracy: 0.7359 - val_loss: 0.5471
Epoch 6/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 391us/step - accuracy: 0.7287 - loss: 0.5520 - val_accuracy: 0.7364 - val_loss: 0.5463
Epoch 7/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 382us/step - accuracy: 0.7297 - loss: 0.5512 - val_accuracy: 0.7341 - val_loss: 0.5473
Epoch 8/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 425us/step - accuracy: 0.7297 - loss: 0

In [35]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - 223us/step - accuracy: 0.7283 - loss: 0.6008
Loss: 0.6007866859436035, Accuracy: 0.7282798886299133


# Attempt 3 Results
** Accuracy:

The accuracy slightly improved compared to previous runs of Attempt 3 but remains consistent at 72.83%.
This suggests that the model is stable and not significantly impacted by random initialization or minor changes in preprocessing.

** Loss:

The loss remained at 0.6008, higher than earlier attempts (e.g., Attempt 1 with 0.5584). This indicates that while the model achieves relatively high accuracy, its predictions may be less precise in some cases.
Impact of Dropping ASK_AMT:

Accuracy remained stable even after dropping ASK_AMT.
The higher loss may indicate that ASK_AMT provided some subtle predictive power for certain instances, but its exclusion did not significantly harm the model's overall performance.

In [50]:
# Export our model to HDF5 file
nn.save("H5/AlphabetSoupCharity_Attempt_3_final.h5")

print("Model saved successfully!")

Model saved successfully!


In [38]:
# Fourth Attempt - Adjusting neurons without adding layers
# Set the random seed
seed = 42
np.random.seed(seed)
tf.random.set_seed(seed)
random.seed(seed)
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
nn = tf.keras.models.Sequential()

# Input layer and first hidden layer
nn.add(tf.keras.layers.Dense(units=120, input_dim=X_train.shape[1], activation='relu'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=80, activation='relu'))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=40, activation='relu'))

# Fourth hidden layer
nn.add(tf.keras.layers.Dense(units=20, activation='relu'))

# Dropout layer
nn.add(tf.keras.layers.Dropout(0.1))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_5 (Dense)                 │ (None, 120)            │         5,160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 80)             │         9,680 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 40)             │         3,240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 20)             │           820 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 20)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 1)              │            21 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 18,921 (73.91 KB)

 Trainable params: 18,921 (73.91 KB)

 Non-trainable params: 0 (0.00 B)

In [40]:
# Compile the model
nn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
# Check the structure of the model
nn.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_5 (Dense)                 │ (None, 120)            │         5,160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 80)             │         9,680 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 40)             │         3,240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 20)             │           820 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 20)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 1)              │            21 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 18,921 (73.91 KB)

 Trainable params: 18,921 (73.91 KB)

 Non-trainable params: 0 (0.00 B)

In [42]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=150, batch_size=32, verbose=1, validation_split=0.2)


Epoch 1/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 1s 492us/step - accuracy: 0.6918 - loss: 0.6036 - val_accuracy: 0.7304 - val_loss: 0.5526
Epoch 2/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 410us/step - accuracy: 0.7247 - loss: 0.5605 - val_accuracy: 0.7357 - val_loss: 0.5498
Epoch 3/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 444us/step - accuracy: 0.7281 - loss: 0.5568 - val_accuracy: 0.7316 - val_loss: 0.5502
Epoch 4/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 410us/step - accuracy: 0.7264 - loss: 0.5557 - val_accuracy: 0.7345 - val_loss: 0.5509
Epoch 5/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 410us/step - accuracy: 0.7292 - loss: 0.5544 - val_accuracy: 0.7353 - val_loss: 0.5497
Epoch 6/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 410us/step - accuracy: 0.7282 - loss: 0.5531 - val_accuracy: 0.7285 - val_loss: 0.5509
Epoch 7/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 416us/step - accuracy: 0.7267 - loss: 0.5515 - val_accuracy: 0.7337 - val_loss: 0.5496
Epoch 8/150
644/644 ━━━━━━━━━━━━━━━━━━━━ 0s 447us/step - accuracy: 0.7285 - loss: 0

In [43]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - 233us/step - accuracy: 0.7277 - loss: 0.5847
Loss: 0.5846846699714661, Accuracy: 0.7276967763900757


# Attempt 4 Results
** Accuracy:

The accuracy is slightly lower than Attempt 3 (72.83%) but similar to Attempt 2 (72.76%).
This indicates that the increased neurons did not provide a significant improvement over earlier attempts.

** Loss:

The loss decreased to 0.5847, lower than Attempt 3 (0.6008) but higher than Attempt 1 (0.5584). This suggests improved prediction precision compared to Attempt 3.
Impact of Neuron Adjustments:

Adjusting the neurons in the fourth attempt did not result in a significant accuracy boost but improved the balance between accuracy and loss.
Attempt 4 demonstrates a reasonable trade-off between prediction precision (low loss) and generalization (accuracy).

In [46]:
# Export our model to HDF5 file
nn.save("H5/AlphabetSoupCharity_Attempt_4.h5")

print("Model saved successfully!")

Model saved successfully!
